src/machine/machined-varlink.c

```diff
@@ -415,32 +417,78 @@ static int list_machine_one(sd_varlink *link, Machine *m, bool more) {
+                { "pid",  SD_JSON_VARIANT_UNSIGNED, sd_json_dispatch_uint32,       offsetof(MachineListQueryParams, pid),          0 },
                 {}
         };
 
-        Manager *m = ASSERT_PTR(userdata);
-        const char *mn = NULL;
         int r;
+        Manager *m = ASSERT_PTR(userdata);
+        Machine *machine = NULL, *pid_machine = NULL;
+        MachineListQueryParams p = {
+                .machine_name = NULL,
+                .pid = -1,
+        };
 
+        assert_cc(sizeof(pid_t) == sizeof(uint32_t));
         assert(parameters);
 
-        r = sd_varlink_dispatch(link, parameters, dispatch_table, &mn);
+        r = sd_varlink_dispatch(link, parameters, dispatch_table, &p);
         if (r != 0)
                 return r;
 
-        if (mn) {
-                if (!hostname_is_valid(mn, /* flags= */ VALID_HOSTNAME_DOT_HOST))
-                        return sd_varlink_error_invalid_parameter_name(link, "name");
+        if (p.machine_name) {
+                r = lookup_machine_by_name(link, m, p.machine_name, &machine);
+                if (r != 0)
+                        return r;
 
-                Machine *machine = hashmap_get(m->machines, mn);
                 if (!machine)
                         return sd_varlink_error(link, "io.systemd.Machine.NoSuchMachine", NULL);
+        }
+
+        if (p.pid >= 0) {
+                if (p.pid == 0) {
+                        int pidfd = sd_varlink_get_peer_pidfd(link);
+                        if (pidfd < 0)
+                                return pidfd;
+
+                        r = pidfd_get_pid(pidfd, &p.pid);
+                        if (r < 0)
+                                return varlink_log_errno(link, r, "Failed to acquire pid of peer: %m");
+                }
+
+                if (p.pid <= 0)
+                        return sd_varlink_error_invalid_parameter_name(link, "pid");
+
+                pid_machine = hashmap_get(m->machine_leaders, PID_TO_PTR(p.pid));
+                if (!pid_machine) {
+                        _cleanup_free_ char *unit = NULL;
+                        r = cg_pid_get_unit(p.pid, &unit);
+                        if (r >= 0)
+                                pid_machine = hashmap_get(m->machine_units, unit);
+                }
+
+                if (!pid_machine)
+                        return sd_varlink_error(link, "io.systemd.Machine.NoSuchMachine", NULL);
+        }
+
+        if (machine && pid_machine) {
+                if (machine != pid_machine)
+                        return sd_varlink_error(link, "io.systemd.Machine.NoSuchMachine", NULL);
 
                 return list_machine_one(link, machine, /* more= */ false);
         }
+        if (machine)
+                return list_machine_one(link, machine, /* more= */ false);
+        if (pid_machine)
+                return list_machine_one(link, pid_machine, /* more= */ false);
```

Comment:
I think we should refuse when the machine found by name and pid are different.
```c
if (p.pid >= 0) {
        Machine *machine;

        r = lookup_machine_by_pid(..., &machine);
        if (r != 0)
                return r;

        if (p.machine_name) {
               /* check found machine name */
        }

        return list_machine_one(link, machine, /* more= */ false);
}

if (p.machine_name) {
        Machine *machine;

        r = lookup_machine_by_pid(..., &machine);
        if (r != 0)
                return r;

        return list_machine_one(link, machine, /* more= */ false);
}
```

Reply:
| I think we should refuse when the machine found by name and pid are different.

This is what the current code does. Per @poettering's suggestion, if both name and pid are specified, the code returns the machine if both match. If any doesn't match, the code returns NoMachine.

However, your code suggests early returns. If pid is specified and there is a match, the machine name is not considered, even if specified in the request. So, I'm confused. Please clarify.

Reply:
Ah yeah, indeed.

So, my suggestion is simply that returns earlier.

Reply:
I'm not sure. I don't think it's a good idea because the result will depend on the order of checks. So, if both "name" and "pid" are present only one of them is respected for non-obvious (for a user) reasons. So, I support Lennart's original suggestion to filter by both attributes when "name" and "pid" specified.